Ar forecasting example

NathanielF opened the pull request:
# Forecasting AR Structural Timeseries models
<!-- Thank you so much for your PR to pymc-examples!

To make the merge process smoother we've provided some links and a checklist below.

We understand that PRs can sometimes be overwhelming, especially as the reviews start coming in.
Please let us know if the reviews are unclear or the recommended next step seems overly demanding,
if you would like help in addressing a reviewer's comments,
or if you have been waiting too long to hear back on your PR. -->

Adding this merge request here to for the open issue: https://github.com/pymc-devs/pymc-examples/issues/450
The notebook has the broad structure of the related blog post and i'm happy to take any feedback or suggestions on streamlining it. I believe it follows the jupyter style advice accurately. 

It seems to be passing the pre-commit checks locally. I found it a bit frustrating to get jupytext to pass. 
<img width="1102" alt="image" src="https://user-images.githubusercontent.com/10588167/197293253-eec03cc7-f789-4cb4-a904-1e29230d73ee.png">

+ [X] Notebook follows style guide https://docs.pymc.io/en/latest/contributing/jupyter_style.html
+ [X] PR description contains a link to the relevant issue: a tracker one for existing notebooks or a proposal one for new notebooks
+ [X] Check the notebook is not excluded from any pre-commit check: https://github.com/pymc-devs/pymc-examples/blob/main/.pre-commit-config.yaml


### Helpful links
* https://github.com/pymc-devs/pymc-examples/blob/main/CONTRIBUTING.md

drbenvincent commented:
Thanks so much for the contribution so far!

Just set the remote checks running. Will try to review over the weekend, but might drift into early next week.

NathanielF commented:
Thanks. Will have a look at that failing check this evening. Think it is failing on the codes poll checks which were skipped for some reason when I ran it locally.

NathanielF commented:
Tried to address the failing codespell check. Seemed to work but not entirely sure. 

<img width="1066" alt="image" src="https://user-images.githubusercontent.com/10588167/197327440-5bdfb816-cd4c-4640-9307-389ce23b9344.png">

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,662 @@
+
+Bayesian structural timeseries models are an interesting way to learn about the structure inherent in any observed timeseries data. It also gives us the ability to project forward the implied predictive distribution granting us another view on forecasting problems. We can treat the learned characteristics of the timeseries data observed to-date as informative about the structure of the unrealised ...
+
+In this notebook we'll see how to fit an predict a range of auto-regressive structural timeseries models and importantly how to predict future observations of the models.
```
```suggestion
In this notebook we'll see how to fit a range of auto-regressive structural timeseries models and importantly how to predict future observations of the models.
```

NathanielF replied:
Ah! I think I get you. Will review the links later tonight and try and adjust to the prediction step everywhere. Any help would be much appreciated! Will add any questions here I find in testing!

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,662 @@
+```{code-cell} ipython3
+az.plot_trace(idata_ar, figsize=(10, 6), kind="rank_vlines");
+```
+
````
I think that it would be valuable to also plot the learnt latent AR variable over time here. You can do something like:
```python
idata_ar.posterior.ar.mean(["chain", "draw"]).plot()
```

or also take advantage of `arviz.hdi`

NathanielF replied:
I've added a plot here but were you thinking i should hit any notes of exposition as well?

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,662 @@
+
+## Prediction Step
+
+The next step works much like generating posterior predictive observations for new data in a GLM model but with one additional complication that the AR process also expects "new" data. Or put another way, we have to tell the model how many prediction steps we want to impute with the model we have just fit. So for the shape handling purposes we have to feed our model the observed data and an appro ...
```
This step is super important but I’m sorry to have to tell you that there is something wrong. This had come up before in a thread in [discourse](https://discourse.pymc.io/t/making-out-of-sample-predictions-with-gaussianrandomwalk/9703/4?u=lucianopaz), but we never found the time to move that into the pymc documentation. I think that your PR is the perfect opportunity to finally add a good explanation on how to do forecasts of time series.
To summarise what was said in the thread, forecasting a time series is not the same as doing predictions with a GLM. The reason these are different is that a time series is conditionally dependent on the past values of the series, whilst the values of a GLM are conditionally independent from each other (given the parameters of the GLM).
The correct way to do a forecast of a time series is to create a new random variable using they learnt AR coefficients, but setting the `init` to a `DiracDelta` of the last entries of the past series. If this isn’t clear, I’m happy to clarify it and provide examples on how this could be done.
@ricardoV94  has a very nice gist that concatenates two random variables, one defined on constant data, and the other on mutable data, which could be a very nice way to do this here.
The other subtlety of forecasting a time series is whether the series is latent or if it was observed. The only real distinction between the two cases is whether the forecast’s init should be defined in terms of the AR random variable, or in terms of its observations directly.

NathanielF replied:
Yes, so i've re-run the notebook with a different pattern of predict step based on the above. Not sure it's exactly what you had in mind, but it seems to work. Base pattern is like this: 

<img width="951" alt="image" src="https://user-images.githubusercontent.com/10588167/197362951-f140d34d-84e2-47c3-9d60-dc0e94cbfe50.png">

The model structure changes here too, so as to emphasise the conditional nature of the forecast. I think this is big improvement on the prior method and really appreciate the pointer, but do let me know if you had something else in mind?

<img width="676" alt="image" src="https://user-images.githubusercontent.com/10588167/197362999-2e2e3a54-42d6-4f9e-8c31-405c2c9c3c90.png">

With similar adjustments made for the trend and seasonal constructs. 

Plotting still roughly works the same way: 

<img width="1117" alt="image" src="https://user-images.githubusercontent.com/10588167/197363034-e75e8efc-9fa6-451c-be75-c5ab40908946.png">

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,662 @@
+plot_fits(idata_ar, idata_preds)
+```
+
+Here we can that although the model converged and ends up with a reasonable fit to the existing the data, and a **plausible  projection** for future values, we have set the prior specification very poorly in allowing an absurdly broad range of due to the kind of compoudning logic of the auto-regressive function. For this reason it's very important to be able to inspect and tailor your model with  ...
````
```suggestion
Here we can see that although the model converged and ends up with a reasonable fit to the existing the data, and a **plausible  projection** for future values, we have set the prior specification very poorly in allowing an absurdly broad range of observations due to the kind of compoudning logic of the auto-regressive function. For this reason it's very important to be able to inspect and tailor your model with prior predictive checks. 
```

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,662 @@
+ax.set_title("AR + Trend + Seasonality");
+```
+
+The key to fitting this model is to understand that we're now passing in synthetic fourier features and that the model now expects a linear combination of those synthetic features even in the prediction step. As such we need to be able to supply those features even out into the future. This fact remains key for any other type of predictive feature we might want to add e.g. day of the week, holida ...
````
I find this paragraph very hard to read. I understand what you want to say, but maybe you could pause a bit more to give a longer explanation of what it is that you’re doing. Imagine that many readers won’t know what a Fourier decomposition is, so maybe you can give some intuition behind (without going into the math though)

NathanielF replied:
I've re-written this paragraph and expanded it out a bit more (hopefully clearer).

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,662 @@
+plot_fits(idata_t_s, preds_t_s)
+```
+
+We can see here how the the model fit again recovers the broad structure and trend of the data, but in addition we have captured the oscillation of the seasonal effect and projected that into the future.
````
```suggestion
We can see here how the model fit again recovers the broad structure and trend of the data, but in addition we have captured the oscillation of the seasonal effect and projected that into the future.
```

lucianopaz commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,662 @@
+```
+
+We can see here how the the model fit again recovers the broad structure and trend of the data, but in addition we have captured the oscillation of the seasonal effect and projected that into the future.
+
````
Maybe you could add some closing remarks here?

NathanielF replied:
Added a closing paragraph too

lucianopaz reviewed (changes requested):
@NathanielF, first of all, thank you for submitting such an awesome example. It touches on a subject that has been completely neglected in the past pymc documentation.
However, there is a crucial subtlety about time series forecasting that was missed in your notebook: the future AR values are conditionally dependent on the past AR values **that had been learnt**. With the approach you did, setting the data of the AR you are sampling the past AR values, and not drawing them from their learnt posterior. The resampled AR will still be conditioned on the learnt coefficients, but the exact AR past values that are in the posterior will be ignored. I mentioned a way to work around this problem [here](https://discourse.pymc.io/t/making-out-of-sample-predictions-with-gaussianrandomwalk/9703/4?u=lucianopaz), and maybe @ricardoV94  can share a gist he wrote that concatenates two random variables to do predictions without sampling what was learnt for the past values.
I would be very happy to help you out in fixing the forecasting, so let me know if my comments were clear or if you need more tips to get this to work.
As I said before and in one of my code comments, this has been lacking in the pymc documentation for a long time, and your contribution is a perfect opportunity to make things right!

NathanielF commented:
Ooops, I didn't mean to remove the request for review from @drbenvincent. 

Also, didn't mean to rush you @lucianopaz. Just wondering is the "request re-review" button is the right etiquette or would you already have seen the changes i've made since you requested them?

lucianopaz reviewed (changes requested):
This looks way better now! I've added a bunch of comments. After you iterate through those, I think that the only things that might be left are a bunch of minor stylistic best practices, and potentially also run `black` and `isort` on the notebook to have it finally ready.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -78,7 +78,12 @@ We'll walk through the model step by step and then generalise the pattern into a
 ## We set up the dictionary to specify size of the AR coefficients in
 ## case we want to vary the AR lags.
-priors = {"coefs": {"mu": 0, "size": 2}, "sigma": 1, "init": {"mu": 8, "sigma": 1, "size": 1}}
+priors = {
+    "coef_0": {"mu": 10, "sigma": 0.1, "size": 2},
+    "coef_1": {"mu": 0.2, "sigma": 0.1, "size": 2},
+    "sigma": 8,
+    "init": {"mu": 9, "sigma": 0.1, "size": 1},
+}
```
Makes sense. I will revert to the coefs pattern.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -104,11 +110,11 @@ with AR:
     ar1 = pm.AR(
         "ar",
-        coefs,
+        at.tensor.as_tensor_variable([coefs_0, coef_1]),
```
Not needed since i'm reverting to coefs and can pass it in directly.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -104,11 +110,11 @@ with AR:
         init_dist=init,
         constant=True,
-        steps=t.shape[0] - (priors["coefs"]["size"] - 1),
+        steps=t.shape[0] - (priors["coef_0"]["size"] - 1),
```
Fair. I'll revert.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,740 @@
+In this notebook we'll see how to fit and predict a range of auto-regressive structural timeseries models and, importantly, how to predict future observations of the models.
+
+```{code-cell} ipython3
+import aesara as at
````
I don't actually need this import anymore once i revert to the coefs pattern

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,740 @@
+    ar1_fut = pm.AR(
+        "ar1_fut",
+        init_dist=pm.DiracDelta.dist(ar1[..., -1]),
+        rho=at.tensor.as_tensor_variable([coefs_0, coef_1]),
```
Again i'll revert to the coefs pattern and the aesera function won't be required.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,740 @@
+    az.plot_ppc(idata_ar, ax=axs[1])
+
+
+plot_fits(idata_ar, idata_preds)
```
This one is tougher. I'll definitely add the legend for the cyan and black lines. I'll also use a mosaic pattern 

AABB
CCCC

to make the plot larger, like so: 
<img width="1130" alt="image" src="https://user-images.githubusercontent.com/10588167/198264265-186b23fb-3956-475c-ba94-9c79dca93936.png">

I'm very hesitant to add a colorbar or more labels to the plot for the colormap. The goal of the colormap is just to convey the graded probability of outcomes in the uncertainty interval and I think it does a good job of that as it is now. I think it would be overkill to add more labels mapping the colors to the range of probability.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,740 @@
+plot_fits(idata_ar, idata_preds)
+```
+
+Here we can see that although the model converged and ends up with a reasonable fit to the existing the data, and a **plausible  projection** for future values. However, we have set the prior specification very poorly in allowing an absurdly broad range of values due to the kind of compoudning logic of the auto-regressive function. For this reason it's very important to be able to inspect and tai ...
````
Good point! Will add an emphasis of this fact.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,740 @@
+    "coef_0": {"mu": 0.2, "sigma": 0.5, "size": 2},
+    "coef_1": {"mu": 0.2, "sigma": 0.03, "size": 2},
+    "alpha": {"mu": -4, "sigma": 0.1},
+    "beta": {"mu": -0.5, "sigma": 0.1},
```
Will adjust these

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,740 @@
+    "coef_0": {"mu": 0.2, "sigma": 0.5, "size": 2},
+    "coef_1": {"mu": 0.2, "sigma": 0.03, "size": 2},
+    "alpha": {"mu": -4, "sigma": 0.1},
+    "beta": {"mu": -0.5, "sigma": 0.1},
```
Will change these too.

NathanielF commented:
Thanks so much for your feedback on this @lucianopaz i will adapt the above discussed and review for grammar and tone. Hopefully push some changes later today.

NathanielF commented:
Ah, sorry @drbenvincent I did the same thing again. No idea why it knocked you out when i requested a review.

NathanielF commented:
Just giving this a slight nudge @lucianopaz , @drbenvincent. Think it's nearly there and it'd be cool if we could get it over the line this week?

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+idata_ar.posterior.ar.mean(["chain", "draw"]).plot(ax=ax, label="Posterior Mean AR level")
+ax.plot(ar1_data, "o", color="black", markersize=2, label="Observed Data")
+ax.legend()
+ax.set_title("Fitted AR process \n and observed data");
```
Very pedantic point, but I don't think we need the spaces around the \n. It acts to make the title off centre.

NathanielF replied:
No problem, will adjust.

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,747 @@
+az.summary(idata_ar, var_names=["~ar"])
+```
+
+```{code-cell} ipython3
````
Maybe just add a short sentence to clarify if a) the fit is bad so we will make structural changes to the model below, or b) if the fit is fine and what step we move on to next.

NathanielF replied:
This is a good point. I'll amend throughout and up the tone of structure helps better modelling. To make my "complications" seem less arbitrary.

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+## Define the time interval for fitting the data
+t_data = list(range(len(ar1_data)))
+## Add the time interval as a mutable coordinate to the model to allow for future predictions
+AR.add_coord("obs_id", t_data, mutable=True)
```
I would be tempted to add `coords` for the two coeffs.

NathanielF replied:
Cool.

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+        init_dist=init,
+        constant=True,
+        steps=t.shape[0] - (priors["coefs"]["size"] - 1),
+    )
```
I'd recommend adding the `obs_id` coeffs here for the `ar` and `likelihood` nodes. That way the graphviz below will render them even better in the same plate as `t` and `y`.

NathanielF replied:
The structure is a little cleaner now: 
<img width="560" alt="image" src="https://user-images.githubusercontent.com/10588167/200553198-b20442ca-2e7d-4e16-b6cb-16ae9cfb0d3d.png">

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,747 @@
+plot_fits(idata_ar, idata_preds)
+```
+
+Here we can see that although the model converged and ends up with a reasonable fit to the existing the data, and a **plausible  projection** for future values. However, we have set the prior specification very poorly in allowing an absurdly broad range of values due to the kind of compounding logic of the auto-regressive function. For this reason it's very important to be able to inspect and tai ...
````
Extra "the" in "reasonable fit to the existing ~the~ data"

NathanielF replied:
Thanks

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+
+Here we can see that although the model converged and ends up with a reasonable fit to the existing the data, and a **plausible  projection** for future values. However, we have set the prior specification very poorly in allowing an absurdly broad range of values due to the kind of compounding logic of the auto-regressive function. For this reason it's very important to be able to inspect and tai ...
+
+Secondly, the mean forecast fails to capture any long lasting structure, quickly dying down to a stable baseline. To account for these kind of short-lived forecasts, we can add more structure to our model, but first, let's complicate the picture. 
```
More of a stylistic suggestion, feel free to ignore. At the moment the wording just suggests increased complexity for the sake of it. Suggest rephrasing to motivate _why_ we might want to add extra model structure... to improve the model in specific ways.

NathanielF replied:
Makes sense.

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+        pass
+
+    t_data = list(range(len(ar_data)))
+    AR.add_coord("obs_id", t_data, mutable=True)
```
Same suggestion about coords and dims as above.

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,747 @@
+plt.suptitle("Prior Predictive Specifications", fontsize=20);
+```
+
+We can see the manner in which the model struggles to capture the trend line. Increasing the variablity of the model will never capture the directional pattern we know to be in the data.
````
"variablity" -> "variability"

NathanielF replied:
done

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+
+### Specifying a Trend Model
+
+We will define a model to account for the trend in our data and combine this trend in a additive model with the autoregressive components. Again the model is much as before, but now we add additional latent features. These are to be combined in a simple additive combination but we can be more creative here if it would suit our model.
```
"trend in a additive" -> "trend in an additive"

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+        pass
+
+    t_data = list(range(len(ar_data)))
+    AR.add_coord("obs_id", t_data, mutable=True)
```
Same suggestion about coords here. I think the most important would be adding `obs_id` as a dim to the `ar` and `likelihood` nodes.

But it would also be good in this model to add a coord for the length 50 variable. At the moment the resulting graphviz is a bit disjointed because it's not matching up nodes in plates.

NathanielF replied:
Done

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,747 @@
+az.summary(idata_trend, var_names=["coefs", "sigma", "alpha", "beta"])
+```
+
+## Complicating the picture further
````
Again, just a stylistic suggestion, but I'd recommend to be more descriptive and positive about modelling more complex structure rather than arbitrarily adding complexity.

NathanielF replied:
I've made it more upbeat with remarks about adding structure to capture aspects of real world data.

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+
+The key to fitting this model is to understand that we're now passing in synthetic fourier features to help account for seasonality effects. This works because (roughly speaking) we're trying to fit a complex oscillating phenomena using a weighted combination of sine and cosine waves. So we add these sine waves and consine waves like we would add any other feature variables in a regression model. ...
+
+However, since we're using this weighted sum to fit the observed data, the model now expects a linear combination of those synthetic features **also** in the prediction step. As such we need to be able to supply those features even out into the future. This fact remains key for any other type of predictive feature we might want to add e.g. day of the week, holiday flag or any other. If a feature  ...
```
Pedantic point, but consider "holiday flag" -> "holiday dummy variable"

NathanielF replied:
Done

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -0,0 +1,747 @@
+
+    ff = ff.to_numpy().T
+    t_data = list(range(len(ar_data)))
+    AR.add_coord("obs_id", t_data, mutable=True)
```
Same suggestion about adding coords and dims. As you can see the resulting graphviz is getting quite gnarly, so by adding these it should act to make this slightly easier to visually parse by the reader.

NathanielF replied:
This is allot better: 
<img width="1265" alt="image" src="https://user-images.githubusercontent.com/10588167/200558463-6f1ac770-6298-44a3-b258-3062638faf38.png">

drbenvincent commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -0,0 +1,747 @@
+plot_fits(idata_t_s, preds_t_s)
+```
+
+We can see here how the the model fit again recovers the broad structure and trend of the data, but in addition we have captured the oscillation of the seasonal effect and projected that into the future.
````
"the the" -> "the"

NathanielF replied:
done.

drbenvincent reviewed (commented):
Hi @NathanielF. Sorry about the delay on the review. I've been bogged down with client work.

It think this is excellent and will make a great addition.

I've added a bunch of comments. Feel free to ask if any clarification is needed.

PS. After these changes I'm happy to approve. That said, I'm not a time series expert, but as long as @lucianopaz is happy then I'm confident we can approve this soon.

NathanielF commented:
Thanks again for your time on this @drbenvincent, I know we're all busy. Happy to wait for @lucianopaz to give the above another look.

I think i've addressed all your comments and his prior comments. In the last commit there i've made the model-diagram neater for all models. I've also stressed that we're adding structural components not for arbitrary reasons but because real-world data tends to have multiple influences... and bayesian structural time-series modelling is one way of capturing theses multi-aspected data generating processes. 

For @lucianopaz 's comments above - the main change was to revert to using the coefs pattern rather than having priors for two individual coefficients separately.

drbenvincent commented:
Just saw this in a quick look now...
In the very final plot, the predicted mean seems not matched up with the shaded regions. Maybe worth a double check.

<img width="689" alt="Screenshot 2022-11-08 at 13 09 04" src="https://user-images.githubusercontent.com/6765047/200573052-9a02cd36-12df-43b4-b8e7-8d4e97bb783c.png">

NathanielF commented:
@drbenvincent do you mean at the tail end of the prediction is seems to come out of phase? Or just about the degree of oscillation? If i change the prior of the beta_fourier terms I get slightly more pronounced oscillation: 

<img width="575" alt="image" src="https://user-images.githubusercontent.com/10588167/200576412-8ba0941b-f9d0-4445-bc15-877ffacef8ee.png">

I wasn't too concerned about it. The point of the plot was to just show that we can recover the seasonality pattern. I think it does that... 

I think it's too much of a rabbit hole to go down, to try and figure out if the percentile color-gradient technique is getting the color map banding exactly right. To my mind the color mapping is there just to suggest that the probabilistic outcomes come with a graded range of plausibility...

drbenvincent commented:
Out was the phase that I noticed. Thought I'd mention it in case or anything obvious.

NathanielF commented:
Right, yeah... honestly not sure why that is happening.

NathanielF commented:
I added more samples to the plot ,extended the prediction period for longer and allowed a wider sigma on the fourier terms.   Still slight phasing visible, but it doesn't seem to be a preface to doom. I don't think it's anything to worry about. 

<img width="574" alt="image" src="https://user-images.githubusercontent.com/10588167/200603199-c3cec61e-a374-4474-9eff-c25beec926e5.png">

lucianopaz reviewed (changes requested):
@NathanielF, thanks again for all of your amazing work! And thanks for the nudge, I had completely lost track of this PR...
I've reviewed it a bit deeper to find out why the forecast plots showed strange behavior. I found 3 issues:

1. The AR distribution you use for forecasting starts from the last time point of the training time. But then you combine it with trend and seasonality that are one time step into the future with respect to it. To fix this, I added an extra coordinate to the model that adds an extra step to the AR so that it has the last time step (used for the init), and all of the time points into the future that must be forecasted.
2. The Fourier features for the forecast with seasonality were not starting from the last time step, but from 0. This added a phase difference that appeared as a sort of discontinuity/inflection between the forecast and the training period.
3. The plots for the cyan lines of the forecast were using the correct future observed time steps, but the shaded areas were using a `linspace`, so they didn't match. Visually, this appeared like a sort of phase lag between the cyan line and the shaded areas.

After addressing these issues locally, the last plot looks like this:

![image](https://user-images.githubusercontent.com/5230109/200802204-07853fdd-d4d4-4e76-a9a6-95331478702e.png)

NathanielF commented:
These are great observations @lucianopaz! Thanks for digging into it. I'll adjust and these and re-push today.

NathanielF commented:
Thanks so much again @lucianopaz that last observation about the plotting function was subtle. I've been looking at the function too long to have seen it! I've adjust the notebook in the manner you suggested and indeed was able to recover a prediction plot with the phasing issues: 

<img width="438" alt="image" src="https://user-images.githubusercontent.com/10588167/200823574-6dd178a4-cb1c-418e-8206-cbbaa3f36d6b.png">

I added a small note about the AR logic in the comments to the code:

<img width="846" alt="image" src="https://user-images.githubusercontent.com/10588167/200823725-fa5607a7-d832-4c7a-9a44-d28860e5ca89.png">

Do you think this is sufficient or should i add anything else?

lucianopaz reviewed (commented):
Thanks @NathanielF. It's almost ready. I just have two very minor nitpicks before approving

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
````diff
@@ -196,7 +198,7 @@ idata_preds
 
 ```{code-cell} ipython3
-def plot_fits(idata_ar, idata_preds):
+def plot_fits(idata_ar, idata_preds, prediction_steps=250):
````
Yep, it can be removed. It is redundant.

NathanielF commented on myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md:
```diff
@@ -226,9 +228,9 @@ def plot_fits(idata_ar, idata_preds):
-        ),
+        az.extract_dataset(idata_ar, group="prior_predictive", num_samples=1000)[
+            "likelihood"
+        ].median(axis=1),
```
This was an accident from when i was experiment with diagnosing the phase shift last night. Reverted now.

NathanielF commented:
Thanks @lucianopaz I fixed those last two issues. It was really great working on this issue with yourself and @drbenvincent I learned a tonne!

lucianopaz reviewed (approved):
Great work @NathanielF! Thank you so much for contributing this!
